### PR TITLE
Pipeline config get and edit APIs

### PR DESCRIPTION
**Pull request by jyotisingh**
We are testing this out currently, and do have a few outstanding issues that need to be fixed, such as handling validations for circular dependencies, quoting etags etc. We would be updating the PR as we have fixes for those.
Please review and provide feedback.

**GET**:  `curl -i 'http://<go-server>:8153/go/api/admin/pipelines/<pipeline-name>' -u 'user:password' -H 'Accept: application/vnd.go.cd.v1+json' -H 'Content-Type: application/json'  -H "If-None-Match: <Etag>"`

**PUT**: `curl -i 'http://<go-server>:8153/go/api/admin/pipelines/<pipeline-name>' -u 'user:password' -H 'Accept: application/vnd.go.cd.v1+json' -H 'Content-Type: application/json' -H 'If-Match: <Etag>' -X PUT -d '{<pipeline json>}'`

**POST**: `curl -i 'http://<go-server>:8153/go/api/admin/pipelines -u 'user:password' -H 'Accept: application/vnd.go.cd.v1+json' -H 'Content-Type: application/json' -X POST -d '{"group": "<group-name>", "pipeline": <pipeline json>}'`

### CODE REVIEW

File: config/config-api/src/com/thoughtworks/go/config/Approval.java [hunk 1]
```diff
@@ -114,21 +130,15 @@ public void validate(ValidationContext validationContext) {
                 return;
             }
             AdminsConfig adminsConfig = group.getAuthorization().getOperationConfig();
-            CruiseConfig cruiseConfig = validationContext.getCruiseConfig();
             for (Admin user : authConfig) {
-                RolesConfig roles = cruiseConfig.server().security().getRoles();
-                if (!userIsAdmin(user, cruiseConfig) && !adminsConfig.has(user, roles.memberRoles(user))) {
+                RolesConfig roles = validationContext.getServerSecurityConfig().getRoles();
+                if (!validationContext.getServerSecurityConfig().isAdmin(user) && !adminsConfig.has(user, roles.memberRoles(user))) {
```
Comment (ketan) [inline config/config-api/src/com/thoughtworks/go/config/Approval.java]: is `validationContext.getServerSecurityConfig()` something that needs to be pulled out of the loop?

File: config/config-api/src/com/thoughtworks/go/config/BasicEnvironmentConfig.java [hunk 2]
```diff
@@ -52,17 +52,17 @@ public BasicEnvironmentConfig(final CaseInsensitiveString name) {
...
+        for (EnvironmentPipelineConfig pipelineRefConfig : this.pipelines) {
+            PipelineConfig pipelineConfig = validationContext.getPipelineConfigByName(pipelineRefConfig.getName());
+            if (pipelineConfig == null)
                 continue;//other rule will error that we reference unknown pipeline
-            if(!cruiseConfig.getConfigRepos().isReferenceAllowed(this.origin,pipelineConfig.getOrigin()))
-                pipelineRefConfig.addError(EnvironmentPipelineConfig.ORIGIN,
-                        String.format("Environment defined in %s cannot reference a pipeline in %s",
-                                this.origin,displayNameFor(pipelineConfig.getOrigin())));
+            if (validationContext.shouldCheckConfigRepo()) {
+                if (!validationContext.getConfigRepos().isReferenceAllowed(this.origin, pipelineConfig.getOrigin()))
```
Comment (ketan) [inline config/config-api/src/com/thoughtworks/go/config/BasicEnvironmentConfig.java]: Pull `validationContext.getConfigRepos()` out of loop?

File: config/config-api/src/com/thoughtworks/go/config/PipelineConfigurationCache.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ************************GO-LICENSE-END
+ ***********************************/
```
Comment (ketan) [inline config/config-api/src/com/thoughtworks/go/config/PipelineConfigurationCache.java]: messed up license?

File: config/config-api/src/com/thoughtworks/go/config/Approval.java (hunk 1 above)
Comment (jyotisingh) [inline config/config-api/src/com/thoughtworks/go/config/Approval.java]: makes sense, will do

File: config/config-api/src/com/thoughtworks/go/config/BasicEnvironmentConfig.java (hunk 2 above)
Comment (jyotisingh) [inline config/config-api/src/com/thoughtworks/go/config/BasicEnvironmentConfig.java]: makes sense, will do

File: config/config-api/src/com/thoughtworks/go/config/PipelineConfigurationCache.java (hunk 3 above)
Comment (jyotisingh) [inline config/config-api/src/com/thoughtworks/go/config/PipelineConfigurationCache.java]: yup :stuck_out_tongue:

**Comment by ketan** [comment]
More comments since the Github does not allow me to add comments on a large commit. Use your judgement to see if some of these can be done in a subsequent request in the interest of merging this quicker –

**`tracking_tool_representer_spec`**

* move `external_tool_hash` into the `external_tool` describe block
* move `mingle_tracking_tool_hash ` into the `mingle` describe block
* missing test for de-serializing with an invalid `type` attribute (see the bottom of `task_representer_spec`)

**`spec_helper.rb`**

* comment out the `backtrace_exclusion_patterns` line, this is mostly useful during development.

**`CatchJsonParseErrors`**

* [convention nazi]please surround method args with brackets in method definitions and invocations. It is cool to drop the brackets for scripty stuff, not for production code.[/convention nazi]

**`routes.rb`**

* please run `rake  --rakefile server/run_rspec_tests.rake exec command='rake routes CONTROLLER=api_v1/admin/pipelines'` – we have effectively defined both singleton and non-singleton restful routes. I *think* there's no need for the `member` block. See [singular resources](http://guides.rubyonrails.org/routing.html#singular-resources) for details. 

```
               Prefix Verb  URI Pattern                          Controller#Action
 apiv1_admin_pipeline GET   /api/admin/pipelines/:name(.:format) api_v1/admin/pipelines#show
                      PATCH /api/admin/pipelines/:name(.:format) api_v1/admin/pipelines#update
                      POST  /api/admin/pipelines/:name(.:format) api_v1/admin/pipelines#create
apiv1_admin_pipelines POST  /api/admin/pipelines(.:format)       api_v1/admin/pipelines#create
                      GET   /api/admin/pipelines/:name(.:format) api_v1/admin/pipelines#show
                      PATCH /api/admin/pipelines/:name(.:format) api_v1/admin/pipelines#update
                      PUT   /api/admin/pipelines/:name(.:format) api_v1/admin/pipelines#update
```

**Pull request merged**